### ISSUE TITLE
ReAdd support for ObserveFinalizeKind

### ISSUE THREAD

**Issue by lionelvillard** [knative/pkg#1668]
I have a use case with 2 controllers, one setting finalizers, and the other one just observing that CRs have been marked for deletion. When the second controller sees the "marked for deletion" events, it updates some external state. The first controller checks this external state before removing finalizers. 

The current reconciler framework does not support this use case anymore (see https://github.com/knative/pkg/pull/2248)
 
/cc @markusthoemmes

**Comment by markusthoemmes** [comment]
Hard time parsing what this means :thinking:.

Delete events are generally not guaranteed to be delivered. If your controller crashes at the same time the delete event is sent and received (or during handling the delete event) the replacement pod for it will not observe the delete event at all.

That's why K8s is doing the "update-with-deletiontimestamp" dance, to guarantee observation, even in edge-cases like this.

**Comment by lionelvillard** [comment]
>  If your controller crashes at the same time the delete event is sent and received (or during handling the delete event) the replacement pod for it will not observe the delete event at all.

This is fine since the goal is to cleanup in-memory state, which is happening when the controller crashes.

TL;DR: I still think there  is a case for observing an object has been marked at deleted without setting a finalizer. `ObserveDeletion` does not work since this function is called when the object has been deleted. 

My use case is I have 2 controllers, one updating resources (e.g. KafkaSource) and the other one just observing resources. When a resource is deleted, I want both controllers to receive the delete event. As it happens the first controller deletes an external resource, which might fails, so a finalizer is set by this controller. The second controller just needs to cleanup in-memory state. There shouldn't be a need to set a finalizer here. 

Bonus: for KafkaSource specifically, the finalizer set by the first controller can't be removed because the operation done in `FinalizeKind` fails due to the in-memory state in the second controller not being cleared up (read: it keeps a connection to an external resource). I know, quite confusing.

**Comment by markusthoemmes** [comment]
So you basically want a coordinated shutdown between N controllers, where the leader is waiting for all other replicas to "ack" that they've observed the finalization and only then will continue?

I suppose we could bring `ObserveFinalizeKind` back for that, but slap huge warnings on it that it's only to be applied in very specific use-cases like this and that `ObserveDeletion` shall be used in any other case.

To take a step back though: Your use-case sounds like an ownership problem. Who owns the resource that gets deleted? It sounds as if there should be an intermediary resource between KafkaSource and X, which the KafkaSource owns and X reconciles the status of. That should be able to give you the very same semantics, I believe.

**Comment by lionelvillard** [comment]
> I suppose we could bring ObserveFinalizeKind back for that, but slap huge warnings on it that it's only to be applied in very specific use-cases like this and that ObserveDeletion shall be used in any other case.

I agree. Will do.

> To take a step back though: Your use-case sounds like an ownership problem. Who owns the resource that gets deleted? It sounds as if there should be an intermediary resource between KafkaSource and X, which the KafkaSource owns and X reconciles the status of. That should be able to give you the very same semantics, I believe.

Yes that's is completely accurate. In my use case, X is a consumer group. We could introduce a new internal ConsumerGroup CRD but that seems like a lot of extra work compare to `ObserveFinalizeKind`. 

Thanks @markusthoemmes, I really appreciate your deep understanding in this matter.

**Comment by lionelvillard** [comment]
You got me thinking more about this and I'm convinced it is worth the effort. I just checked it's technically feasible.

thanks for you help, really appreciated!
